### ISSUE TITLE
Compiler panicking due to unsuccessful type parameter resolution

### ISSUE THREAD

**Issue by Saxarok** [rust-lang/rust#98031]
### Code

```Rust
use std::marker::PhantomData;

fn test<T>() {}

struct Foo<T>(PhantomData<T>);

impl<T> Foo<T> {
    // Can be any type
    const FOO_BAR: () = ();

    fn bar() {
        test();
        
        // If you remove this line compiler throws an error
        // error[E0282]: type annotations needed
        Self::FOO_BAR;
    }
}
```

### Meta

`rustc --version --verbose`:
```
rustc 1.63.0-nightly (ec55c6130 2022-06-10)
binary: rustc
commit-hash: ec55c61305eaf385fc1b93ac9a78284b4d887fe5
commit-date: 2022-06-10
host: x86_64-pc-windows-msvc
release: 1.63.0-nightly
LLVM version: 14.0.5
```

### Error output

```
   Compiling crash-test v0.1.0 (C:\Users\Saxarok\Desktop\dev\crash-test)
error: internal compiler error: compiler\rustc_infer\src\infer\error_reporting\need_type_info.rs:856:33: unexpected path: def=Foo substs=[_] path=Path { span: src\main.rs:14:9: 14:13 (#0), res: SelfTy { trait_: None, alias_to: Some((DefId(0:12 ~ crash_test[4dc2]::{impl#0}), false)) }, segments: [PathSegment { ident: Self#0, hir_id: Some(HirId { owner: DefId(0:15 ~ crash_test[4dc2]::{impl#0}::bar), local_id: 5 }), res: Some(SelfTy { trait_: None, alias_to: Some((DefId(0:12 ~ crash_test[4dc2]::{impl#0}), false)) }), args: None, infer_args: true }] }

thread 'rustc' panicked at 'Box<dyn Any>', /rustc/ec55c61305eaf385fc1b93ac9a78284b4d887fe5\compiler\rustc_errors\src\lib.rs:1339:9
stack backtrace:
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.63.0-nightly (ec55c6130 2022-06-10) running on x86_64-pc-windows-msvc

note: compiler flags: --crate-type bin -C embed-bitcode=no -C debuginfo=2 -C incremental

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [typeck] type-checking `<impl at src\main.rs:7:1: 16:2>::bar`
#1 [typeck_item_bodies] type-checking all item bodies
#2 [analysis] running analysis passes on this crate
end of query stack
error: could not compile `crash-test`
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
   0:     0x7ff882889d1f - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h5878974d3101c2c1
   1:     0x7ff8828c4d6a - core::fmt::write::h58b91005b62f62cb
   2:     0x7ff88287c119 - <std::io::IoSlice as core::fmt::Debug>::fmt::h47ce2c587850b386
   3:     0x7ff88288d60b - std::panicking::default_hook::h1cf6e1ccf8670b6e
   4:     0x7ff88288d28b - std::panicking::default_hook::h1cf6e1ccf8670b6e
   5:     0x7ff8683db786 - rustc_driver[a11d5e36637d6ef8]::pretty::print_after_hir_lowering
   6:     0x7ff88288ddb2 - std::panicking::rust_panic_with_hook::h8d89018be48b3be7
   7:     0x7ff86c9b1565 - <rustc_middle[fa74f14430eff12]::mir::interpret::allocation::ConstAllocation as rustc_middle[fa74f14430eff12]::ty::context::Lift>::lift_to_tcx
   8:     0x7ff86c9aed79 - <rustc_middle[fa74f14430eff12]::mir::interpret::allocation::ConstAllocation as rustc_middle[fa74f14430eff12]::ty::context::Lift>::lift_to_tcx
   9:     0x7ff86cedbbc9 - rustc_middle[fa74f14430eff12]::util::bug::bug_fmt
  10:     0x7ff86c947d59 - <rustc_middle[fa74f14430eff12]::mir::GeneratorInfo as core[23edea91be4138c2]::fmt::Debug>::fmt
  11:     0x7ff86c9464f3 - <rustc_middle[fa74f14430eff12]::mir::GeneratorInfo as core[23edea91be4138c2]::fmt::Debug>::fmt
  12:     0x7ff86c945ed2 - <rustc_middle[fa74f14430eff12]::mir::GeneratorInfo as core[23edea91be4138c2]::fmt::Debug>::fmt
  13:     0x7ff86c93c439 - <rustc_middle[fa74f14430eff12]::ty::consts::valtree::ValTree>::zst
  14:     0x7ff86c93e9dc - <rustc_middle[fa74f14430eff12]::ty::generics::GenericPredicates>::instantiate_identity
  15:     0x7ff86ced8c48 - rustc_middle[fa74f14430eff12]::util::bug::bug_fmt
  16:     0x7ff86c802759 - <rustc_infer[debc1fc2b0c2d569]::infer::error_reporting::need_type_info::TypeAnnotationNeeded as core[23edea91be4138c2]::convert::Into<rustc_errors[233cc4104dd9fe30]::diagnostic::DiagnosticId>>::into
  17:     0x7ff86c802eb1 - <rustc_infer[debc1fc2b0c2d569]::infer::error_reporting::need_type_info::FindInferSourceVisitor as rustc_hir[2565c86d9e92d930]::intravisit::Visitor>::visit_expr
  18:     0x7ff86c85270d - <rustc_infer[debc1fc2b0c2d569]::infer::lexical_region_resolve::RegionResolutionError as core[23edea91be4138c2]::fmt::Debug>::fmt
  19:     0x7ff86c802c64 - <rustc_infer[debc1fc2b0c2d569]::infer::error_reporting::need_type_info::FindInferSourceVisitor as rustc_hir[2565c86d9e92d930]::intravisit::Visitor>::visit_expr
  20:     0x7ff86c774421 - <rustc_infer[debc1fc2b0c2d569]::infer::InferCtxt>::emit_inference_failure_err
  21:     0x7ff86c71476c - <rustc_infer[debc1fc2b0c2d569]::infer::InferCtxt as rustc_trait_selection[87fd71c093326866]::traits::error_reporting::InferCtxtPrivExt>::maybe_report_ambiguity
  22:     0x7ff86c700be5 - <rustc_infer[debc1fc2b0c2d569]::infer::InferCtxt as rustc_trait_selection[87fd71c093326866]::traits::error_reporting::InferCtxtExt>::report_fulfillment_errors
  23:     0x7ff86ad92374 - <rustc_typeck[da438188b5776efb]::check::fn_ctxt::FnCtxt>::apply_adjustments
  24:     0x7ff86aeaf565 - <rustc_infer[debc1fc2b0c2d569]::infer::outlives::env::OutlivesEnvironment as rustc_typeck[da438188b5776efb]::check::regionck::OutlivesEnvironmentExt>::add_implied_bounds
  25:     0x7ff86adf43b5 - rustc_typeck[da438188b5776efb]::check::provide
  26:     0x7ff86b9ad2b8 - <&[rustc_span[bc679ff3d8b5d5b9]::def_id::LocalDefId] as rustc_serialize[3bc617203079c8a5]::serialize::Decodable<rustc_query_impl[72833c6ed9f636d4]::on_disk_cache::CacheDecoder>>::decode
  27:     0x7ff86ba294a2 - <&[rustc_span[bc679ff3d8b5d5b9]::symbol::Ident] as rustc_serialize[3bc617203079c8a5]::serialize::Decodable<rustc_query_impl[72833c6ed9f636d4]::on_disk_cache::CacheDecoder>>::decode
  28:     0x7ff86bb691d0 - <&[rustc_span[bc679ff3d8b5d5b9]::symbol::Ident] as rustc_serialize[3bc617203079c8a5]::serialize::Decodable<rustc_query_impl[72833c6ed9f636d4]::on_disk_cache::CacheDecoder>>::decode
  29:     0x7ff86bcacd18 - <&[rustc_span[bc679ff3d8b5d5b9]::symbol::Ident] as rustc_serialize[3bc617203079c8a5]::serialize::Decodable<rustc_query_impl[72833c6ed9f636d4]::on_disk_cache::CacheDecoder>>::decode
  30:     0x7ff86ac55b81 - <rustc_typeck[da438188b5776efb]::outlives::explicit::ExplicitPredicatesMap as core[23edea91be4138c2]::fmt::Debug>::fmt
  31:     0x7ff86adfa493 - <rustc_typeck[da438188b5776efb]::check::MaybeInProgressTables>::borrow_mut
  32:     0x7ff86b9af0ae - <&[rustc_span[bc679ff3d8b5d5b9]::def_id::LocalDefId] as rustc_serialize[3bc617203079c8a5]::serialize::Decodable<rustc_query_impl[72833c6ed9f636d4]::on_disk_cache::CacheDecoder>>::decode
  33:     0x7ff86ba5190f - <&[rustc_span[bc679ff3d8b5d5b9]::symbol::Ident] as rustc_serialize[3bc617203079c8a5]::serialize::Decodable<rustc_query_impl[72833c6ed9f636d4]::on_disk_cache::CacheDecoder>>::decode
  34:     0x7ff86bbff81d - <&[rustc_span[bc679ff3d8b5d5b9]::symbol::Ident] as rustc_serialize[3bc617203079c8a5]::serialize::Decodable<rustc_query_impl[72833c6ed9f636d4]::on_disk_cache::CacheDecoder>>::decode
  35:     0x7ff86bc77c09 - <&[rustc_span[bc679ff3d8b5d5b9]::symbol::Ident] as rustc_serialize[3bc617203079c8a5]::serialize::Decodable<rustc_query_impl[72833c6ed9f636d4]::on_disk_cache::CacheDecoder>>::decode
  36:     0x7ff86ae4ff1f - <rustc_typeck[da438188b5776efb]::structured_errors::wrong_number_of_generic_args::GenericArgsInfo as core[23edea91be4138c2]::fmt::Debug>::fmt
  37:     0x7ff86ac51626 - rustc_typeck[da438188b5776efb]::check_crate
  38:     0x7ff8684c8c1f - rustc_interface[c44e93edb349834f]::passes::analysis
  39:     0x7ff86b9aecde - <&[rustc_span[bc679ff3d8b5d5b9]::def_id::LocalDefId] as rustc_serialize[3bc617203079c8a5]::serialize::Decodable<rustc_query_impl[72833c6ed9f636d4]::on_disk_cache::CacheDecoder>>::decode
  40:     0x7ff86ba489ef - <&[rustc_span[bc679ff3d8b5d5b9]::symbol::Ident] as rustc_serialize[3bc617203079c8a5]::serialize::Decodable<rustc_query_impl[72833c6ed9f636d4]::on_disk_cache::CacheDecoder>>::decode
  41:     0x7ff86bbe9954 - <&[rustc_span[bc679ff3d8b5d5b9]::symbol::Ident] as rustc_serialize[3bc617203079c8a5]::serialize::Decodable<rustc_query_impl[72833c6ed9f636d4]::on_disk_cache::CacheDecoder>>::decode
  42:     0x7ff86bcad126 - <&[rustc_span[bc679ff3d8b5d5b9]::symbol::Ident] as rustc_serialize[3bc617203079c8a5]::serialize::Decodable<rustc_query_impl[72833c6ed9f636d4]::on_disk_cache::CacheDecoder>>::decode
  43:     0x7ff86838fbbf - <rustc_driver[a11d5e36637d6ef8]::args::Error as core[23edea91be4138c2]::fmt::Debug>::fmt
  44:     0x7ff868367210 - <rustc_lint[6c4a321fe5b8ec79]::BuiltinCombinedModuleLateLintPass as rustc_lint[6c4a321fe5b8ec79]::passes::LateLintPass>::check_mod_post
  45:     0x7ff8683d2215 - <rustc_middle[fa74f14430eff12]::ty::SymbolName as core[23edea91be4138c2]::fmt::Debug>::fmt
  46:     0x7ff86836846d - <rustc_lint[6c4a321fe5b8ec79]::BuiltinCombinedModuleLateLintPass as rustc_lint[6c4a321fe5b8ec79]::passes::LateLintPass>::check_mod_post
  47:     0x7ff8683cee76 - <rustc_metadata[7439ebc3b5fbc24c]::rmeta::encoder::EncodeContext as rustc_serialize[3bc617203079c8a5]::serialize::Encoder>::emit_raw_bytes
  48:     0x7ff868392398 - <rustc_driver[a11d5e36637d6ef8]::args::Error as core[23edea91be4138c2]::fmt::Debug>::fmt
  49:     0x7ff88289ef5c - std::sys::windows::thread::Thread::new::h40e1d95cdd0a661f
  50:     0x7ff922967034 - BaseThreadInitThunk
  51:     0x7ff923e02651 - RtlUserThreadStart
```

</p>
</details>

**Comment by matthiaskrgr** [comment]
should be fixed with `rustc 1.63.0-nightly (99930ac7f 2022-06-11)`

**Comment by Saxarok** [comment]
Yes, you're right.
